src/Arcus.Security.Core/Providers/MutatedSecretNameCachedSecretProvider.cs

```diff
@@ -0,0 +1,139 @@
+        /// </summary>
+        /// <param name="secretName">The name of the secret key</param>
+        /// <param name="ignoreCache">Indicates if the cache should be used or skipped</param>
+        /// <returns>Returns a <see cref="Task{TResult}"/> that contains the secret key</returns>
+        /// <exception cref="ArgumentException">The name must not be empty</exception>
+        /// <exception cref="ArgumentNullException">The name must not be null</exception>
+        /// <exception cref="SecretNotFoundException">The secret was not found, using the given name</exception>
+        public async Task<Secret> GetSecretAsync(string secretName, bool ignoreCache)
+        {
+            Guard.NotNullOrWhitespace(secretName, nameof(secretName), "Requires a non-blank secret name when mutating secret names");
+
+            string mutatedSecretName = MutateSecretName(secretName);
+            Task<Secret> secretAsync = _implementation.GetSecretAsync(mutatedSecretName, ignoreCache);
+
+            if (secretAsync is null)
+            {
+                throw new SecretNotFoundException(mutatedSecretName);
+            }
+
+            try
+            {
+                return await secretAsync;
+            }
+            catch (Exception exception)
+            {
+                Logger.LogError(
+                    exception, "Failure during retrieving secret '{MutatedSecretName}' that was mutated from '{OriginalSecretName}'", mutatedSecretName, secretName);
+
+                throw new SecretNotFoundException(mutatedSecretName, exception);
+            }
+        }
+
+        /// <summary>
+        /// Removes the secret with the given <paramref name="secretName"/> from the cache;
+        /// so the next time <see cref="ISecretProvider.GetSecretAsync(string)"/> is called, a new version of the secret will be added back to the cache.
+        /// </summary>
+        /// <param name="secretName">The name of the secret that should be removed from the cache.</param>
+        public async Task InvalidateSecretAsync(string secretName)
+        {
+            Guard.NotNullOrWhitespace(secretName, nameof(secretName), "Requires a non-blank secret name when mutating secret names");
+
+            string mutatedSecretName = MutateSecretName(secretName);
+            Task invalidateSecretAsync = _implementation.InvalidateSecretAsync(mutatedSecretName);
+
+            if (invalidateSecretAsync is null)
+            {
+                throw new SecretNotFoundException(mutatedSecretName);
+            }
+
+            try
+            {
+                await invalidateSecretAsync;
+            }
+            catch (Exception exception)
+            {
+                Logger.LogError(
+                    exception, "Failure during invalidating secret '{MutatedSecretName}' that was mutated from '{OriginalSecretName}'", mutatedSecretName, secretName);
+
+                throw new SecretNotFoundException(mutatedSecretName, exception);
```

Comment:
Is this correct? Who says it was because the secret was not found?

Reply:
Changed to `InvalidOperationException` exception type and moved to central method. Thanks! Resolving.